nexus/mgs-updates/src/host_phase1_updater.rs

```diff
@@ -176,3 +309,388 @@ impl SpComponentUpdater for HostPhase1Updater {
+            return Err(PrecheckError::WrongActiveArtifact {
+                kind: ArtifactKind::HOST_PHASE_1,
+                expected: *expected_active_phase_1_hash,
+                found: current_active_slot_hash,
+            });
+        }
+
+        // For the same reason, check that the artifact hash in the inactive
+        // slot matches what we expect to find.
+        let expected_inactive_slot = expected_active_phase_1_slot.toggled();
+        let found_inactive_artifact = self
+            .precheck_fetch_phase_1(
+                mgs_clients,
+                update.sp_type,
+                update.slot_id,
+                expected_inactive_slot,
+                log,
+            )
+            .await?;
+        debug!(
+            log, "found inactive slot phase 1 artifact";
+            "hash" => %found_inactive_artifact,
+        );
+
+        if found_inactive_artifact == *expected_inactive_phase_1_hash {
+            // Everything looks good as far as phase 1 preconditions are
+            // concerned; also confirm that our phase 2 preconditions are
+            // satisfied by contacting sled-agent.
+            self.precheck_phase_2_via_sled_agent(log).await?;
+
+            Ok(PrecheckStatus::ReadyForUpdate)
+        } else {
+            Err(PrecheckError::WrongInactiveArtifact {
+                kind: ArtifactKind::HOST_PHASE_1,
+                expected: *expected_inactive_phase_1_hash,
+                found: found_inactive_artifact,
+            })
+        }
+    }
+
+    async fn precheck_fetch_phase_1(
+        &self,
+        mgs_clients: &mut MgsClients,
+        sp_type: SpType,
+        sp_slot: u16,
+        target_slot: M2Slot,
+        log: &Logger,
+    ) -> Result<ArtifactHash, PrecheckError> {
+        match mgs_clients
+            .try_all_serially(log, move |mgs_client| async move {
+                match mgs_client
+                    .host_phase_1_flash_hash_calculate_with_timeout(
+                        sp_type,
+                        sp_slot,
+                        target_slot.to_mgs_firmware_slot(),
+                        PHASE_1_HASHING_TIMEOUT,
+                    )
+                    .await
+                {
+                    // The return types here are a little weird;
+                    // `try_all_serially()` requires us to return a `Result<T,
+                    // GatewayClientError>`, but
+                    // `host_phase_1_flash_hash_calculate_with_timeout()`
+                    // returns a `HostPhase1HashError`; its `RequestError`
+                    // variant _contains_ a `GatewayClientError`. We convert
+                    // that specific variant into its `GatewayClientError`, and
+                    // return a `Result<Result<_, HostPhase1HashError>,
+                    // GatewayClientError>. We unpack the inner result in a
+                    // `match` after `try_all_serially()`.
```

Comment:
This is okay, but it does make me wonder if this logic belongs outside the call to `try_all_serially` (which I think was envisioned as a thin wrapper around individual MGS client calls, hence no need to return other kinds of errors).